src/plots/cartesian/axis_defaults.js

```diff
@@ -142,6 +138,11 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, coerce,
         delete containerOut.zerolinewidth;
     }
 
+    // fill in categories
+    containerOut._initialCategories = axType === 'category' ?
+        orderedCategories(letter, containerOut.categoryorder, containerOut.categoryarray, options.data) :
+        [];
+
     return containerOut;
```

Comment:
It would deserve more than a block comment in my opinion. 

We should try tackling https://github.com/plotly/plotly.js/issues/66 soon.